### ISSUE TITLE
Investigate plugin context menu

### ISSUE THREAD

**Issue by abique** [free-audio/clap#29]

**Comment by baconpaul** [comment]
yeah vst3 let you do that but it was painful (remember the host menu with a ref count, grab the actions, rebuild, release on trigger, etc...)

**Comment by Bremmers** [comment]
The simplest form would be something like this:

typedef struct clap_host_contextmenu {
 uint32_t (*count)(const clap_host_t *host, clap_id param_id);

 bool (*gettitle)(const clap_host_t *host, clap_id param_id, uint32_t index, char *display, uint32_t size);

 void (*clicked)(const clap_host_t *host, clap_id param_id, uint32_t index);
}

The plugin asks for the number of menu items for the param, gets the titles and adds them to its menu. It calls clicked if an item is clicked.

But it probably should support sub-menus too.

**Comment by baconpaul** [comment]
sub menus
menu titles
accelerator keys
checkmarks
active/inactive state

all of those things at least

**Comment by baconpaul** [comment]
maybe not accelerator keys

**Comment by Bremmers** [comment]
Yeah, accelerator keys can conflict with the plugin itself.

**Comment by Bremmers** [comment]
This can do it I think.
count() could optionally be removed (you can call getitem() until it returns false instead)



// the plugin calls count(.., .., CLAP_INVALID_ID) to get the toplevel menu item count
// if count > 0 it fetches the items using getitem(.., .., CLAP_INVALID_ID, ..)
// items with CLAP_MENU_FLAG_HASSUBMENU in flags have a submenu. The plugin calls count(.., .., menu_id) etc.
// submenu items can have a submenu again etc.

// clicked() is called on clicking a menu item
// clicked() isn't called for menu items that have a submenu

enum {
  CLAP_MENU_FLAG_HASSUBMENU = 1 << 0,
  CLAP_MENU_FLAG_DISABLED = 1 << 1,
  CLAP_MENU_FLAG_CHECKED = 1 << 2,
}

typedef struct clap_menuitem {
   char    name[CLAP_NAME_SIZE];
   uint32_t flags;   
   clap_id menu_id;
}

typedef struct clap_host_contextmenu {
 uint32_t (*count)(const clap_host_t *host, clap_id param_id, clap_id menu_id);

 bool (*getitem)(const clap_host_t *host, clap_id param_id, clap_id menu_id, uint32_t index, clap_menuitem *menuitem);

 void (*clicked)(const clap_host_t *host, clap_id param_id, clap_id menu_id);
}

**Comment by Bremmers** [comment]
> Clicked would also need to have an uint32_t index,
Each menu item has it's own menu_id. 'index' can't be used because menus can be nested (a submenu in a submenu).
Typically, a host that uses a simple one-dimensional menu will make menu_id equal to index for all items.

> I'd define a separate static const CLAP_CONSTEXPR clap_id CLAP_EXT_CONTEXT_MENU_ROOT = 0; for the root menu. Having an invalid menu be the top level menu is a bit counter intuitive.

Agreed. 0 is probably not a good idea as the host would probably want to use that (see above)

**Comment by Bremmers** [comment]
// the plugin calls count(.., .., CLAP_EXT_CONTEXT_MENU_ROOT) to get the toplevel menu item count
// if count > 0 it fetches the items using getitem(.., .., CLAP_EXT_CONTEXT_MENU_ROOT, ..)

// items with CLAP_MENU_FLAG_HASSUBMENU in flags have a submenu. The plugin calls count(.., .., menu_id) etc.
// submenu items can have a submenu again etc.
// a plugin that can't display submenus shouldn't display the parent menu item (with CLAP_MENU_FLAG_HASSUBMENU in flags) either

static CLAP_CONSTEXPR const char CLAP_EXT_CONTEXT_MENU[] = "clap.context-menu.draft/0";

static const CLAP_CONSTEXPR clap_id CLAP_EXT_CONTEXT_MENU_ROOT = CLAP_INVALID_ID - 1;

enum {
CLAP_MENU_FLAG_HASSUBMENU = 1 << 0,
CLAP_MENU_FLAG_DEACTIVATED = 1 << 1,
CLAP_MENU_FLAG_CHECKED = 1 << 2,
CLAP_MENU_FLAG_ISSEPARATOR = 1 << 3,
}

typedef struct clap_menuitem {
char name[CLAP_NAME_SIZE];
uint32_t flags;
clap_id menu_id;
}

typedef struct clap_host_context_menu {
//get number of menu items
uint32_t (*count)(const clap_host_t *host, clap_id param_id, clap_id menu_id);

// get menu item description.
// Each item has a unique menu_id which the plugin can pass to clicked()
bool (*getitem)(const clap_host_t *host, clap_id param_id, clap_id menu_id, uint32_t index, clap_menuitem *menuitem);

// called on clicking a menu item
// clicked() isn't called for menu items that have CLAP_MENU_FLAG_HASSUBMENU, CLAP_MENU_FLAG_DEACTIVATED or CLAP_MENU_FLAG_ISSEPARATOR in flags
void (*clicked)(const clap_host_t *host, clap_id param_id, clap_id menu_id);
}

**Comment by abique** [comment]
@Bremmers what about implementing this extension on a branch and making a pull request? ;-)

**Comment by Bremmers** [comment]
I'm still hoping someone who knows how to do that turns this into proper-looking C and makes a PR :-)

**Comment by baconpaul** [comment]
@Bremmers top GitHub tip. If you include code in triple back quotes followed by cpp you get formatted code

```cpp
int main() { 
   return std::max(2,3); 
}
```

<img width="327" alt="Screen Shot 2022-06-23 at 7 18 26 AM" src="https://user-images.githubusercontent.com/13387561/175286888-9d0c3a3b-7193-4c7e-8ab9-203f5274f1d4.png">

**Comment by baconpaul** [comment]
Anyway I really want this feature also. Happy to type up or review, but may not get to it today.

**Comment by Bremmers** [comment]
Wouldn't that cause all kinds of UI framework problems, especially on Linux?
Z order and what not.

**Comment by Bremmers** [comment]
Ah, that works with fabfilter indeed. Didn't know that.

So do you think we should just add a popup function to clap_host_context_menu? Or perhaps there should be two host interfaces, one for plugins that have their own menus, and one for plugins that don't.

Could there be a situation where the plugin UI has become invisible but the menu is still visible? The host would have to hide the menu in the gui.h 'hide' functions perhaps.

**Comment by robbert-vdh** [comment]
Not quite sure. Maybe splitting it up is indeed a better idea, since a host may be able to provide some context menu items but it doesn't have its own way of popping up a context menu (and having some of these functions always return false would also be strange behavior).

> Could there be a situation where the plugin UI has become invisible but the menu is still visible?

The host would be the one hiding the plugin's UI (and if the window was closed through the window manager, it would already handle the event for that). So I don't think that would cause any issues, the host would just need to close the context menu in those places. And if they don't, then it would probably only be a cosmetic problem with the host that doesn't affect any functionality.

**Comment by baconpaul** [comment]
but also: is popup sync or async? I presume async.

**Comment by robbert-vdh** [comment]
> To be a menu provider you need both.

But then how would you prevent the host from simply only implementing one of the two ways to use the interface and just returning false/0 for the other one? I could see that happening.

> but also: is popup sync or async? I presume async.

I'm not sure yet. I'd also presume async would be fine, so I'm trying to think of situations where that could be problematic.

**Comment by Bremmers** [comment]
I'm wondering if the 'mouse pointer position' is always available. It could be a longtap on a touchscreen, or a pen. In theory these devices wouldn't have to affect the mouse pointer position using the Windows API, in reality they probably will.

So perhaps the plugin should set the mouse pointer position if it uses other input devices.

**Comment by robbert-vdh** [comment]
If the location is really not available (but for those long tap like things it should be), then the host can also just open the context menu in the center of the screen. Or they probably have some way to get the position from the touch API themselves. The potential problem with passing coordinates here is that you need to take HiDPI into account and you also need to translate window coordinates to global coordinates. So if you don't need to do these things, there are two fewer ways for hosts to make mistakes here

**Comment by Bremmers** [comment]
```cpp
// clap_host_plugin_context_menu lets a plugin integrate host menu items in its own menus

// the plugin calls count(.., .., CLAP_EXT_CONTEXT_MENU_ROOT) to get the toplevel menu item count
// if count > 0 it fetches the items using getitem(.., .., CLAP_EXT_CONTEXT_MENU_ROOT, ..)

// items with CLAP_MENU_FLAG_HASSUBMENU in flags have a submenu. The plugin calls count(.., .., menu_id) etc.
// submenu items can have a submenu again etc.
// a plugin that can't display submenus shouldn't display the parent menu item (with CLAP_MENU_FLAG_HASSUBMENU in flags) either

static CLAP_CONSTEXPR const char CLAP_EXT_CONTEXT_MENU[] = "clap.context-menu.draft/0";

static const CLAP_CONSTEXPR clap_id CLAP_EXT_CONTEXT_MENU_ROOT = CLAP_INVALID_ID - 1;

enum {
  CLAP_MENU_FLAG_HASSUBMENU = 1 << 0,
  CLAP_MENU_FLAG_DEACTIVATED = 1 << 1,
  CLAP_MENU_FLAG_CHECKED = 1 << 2,
  CLAP_MENU_FLAG_ISSEPARATOR = 1 << 3,
}

typedef struct clap_menuitem {
  char name[CLAP_NAME_SIZE];
  uint32_t flags;
  clap_id menu_id;
}

typedef struct clap_host_plugin_context_menu {
  //get number of menu items
  uint32_t (*count)(const clap_host_t *host, clap_id param_id, clap_id menu_id);

  // get menu item description.
  // Each item has a unique menu_id which the plugin can pass to clicked()
  bool (*getitem)(const clap_host_t *host, clap_id param_id, clap_id menu_id, uint32_t index, clap_menuitem *menuitem);

  // called on clicking a menu item
  // clicked() isn't called for menu items that have CLAP_MENU_FLAG_HASSUBMENU, CLAP_MENU_FLAG_DEACTIVATED or CLAP_MENU_FLAG_ISSEPARATOR in flags
  void (*clicked)(const clap_host_t *host, clap_id param_id, clap_id menu_id);
}

// clap_host_host_context_menu.popup can be called by plugins that don't have their own menus
// the menu should appear at the current mouse pointer position
typedef struct clap_host_host_context_menu {
  void (*popup)(const clap_host_t *host, clap_id param_id);
}
```

**Comment by robbert-vdh** [comment]
Yes, something like that. With the `clap_host_plugin_context_menu` and `clap_host_host_context_menu` extensions having their own extension IDs, of course.

**Comment by robbert-vdh** [comment]
Yes it would need to be two extensions.

**Comment by baconpaul** [comment]
Thinking like a host developer: if it is one interface chance of implementing both is way higher

**Comment by Bremmers** [comment]
Thinking like Alexandre: it probably should be one interface ;-)
That's what other extensions tend to do I think.

**Comment by abique** [comment]
Sorry to jump late into the game.
1. Can a host integrate plugin menu as well? Would the extension be perfectly symmetric for both host and plugin?
2. Do those menu have a context? (global, parameter, modulation, ...)
3. Can they have complex menu entries? (check box and checked state or radio button, icon, ...)?
4. What would you add in those menu (examples)

Depending on the feature requirements, maybe one could consider a json description of the menu items.
Another idea would be to pass a menu builder:

```C
struct clap_menu_builder {
   void add_entry(icon, label, entry_id);
   void add_separator();
   void begin_submenu(icon, label);
   void end_submenu();
};

struct clap_host_context_menu {
   bool create(context, builder);
   void clicked(entry_id);
};
```

Here it might be hard for the host to connect the entry_id back to an action, maybe entry_id could be replaced by an arbitrary blob of bytes that the plugin would copy?

All of this is just suggestions.

**Comment by abique** [comment]
One thing I'm wondering, is the host could inject:
- delete automations
- delete modulation mappings
But the plugin could also inject those option and rely upon `host_params->clear(references)`, and then we'd end up with duplicate menu entries.

For example, the plugin could get a list of available modulation sources from the host, and then add a lot of functionality related to modulations in the context menu without any help from the host.

One other thing is the consistency:
- when using the same plugin in multiple daws
- when using multiple plugins in the same daw

**Comment by Bremmers** [comment]
4) actually all I want to add is "Open automation editor" :-)

Later...

**Comment by baconpaul** [comment]
The plugin menu to the host is an interesting idea but I think the host is always “open” so like if the surge gui is closed you don’t want to show code which would launch an editor callback from bitwig. I wonder if I would implement that

i really think of these as letting the plugin workflow more easily interact with automation in the daw and that’s it - and think a simple api is more likely to get adopted both sides. Agree active headers submenus and checks are a minimum. Icons are super hard.

**Comment by Bremmers** [comment]
Perhaps we should define some predefined standard actions, like 'open automation editor' and 'delete all automation' that are super easy to implement.
In addition there can be custom actions that go all the way, and they're probably only implemented by fancy hosts and synths.

Having predefined standard actions also means the other side knows what it is, so it can prevent duplicates.

Context menus are in the 'nice to have' category for plugin developers, and perhaps not even that. If it isn't super easy they won't implement it.

**Comment by Bremmers** [comment]
> 2\. Do those menu have a context? (global, parameter, modulation, ...)
I think they're always associated with a parameter?

**Comment by baconpaul** [comment]
> > 2. Do those menu have a context? (global, parameter, modulation, ...)
> 
> I think they're always associated with a parameter?

Yes

**Comment by robbert-vdh** [comment]
> Can a host integrate plugin menu as well? Would the extension be perfectly symmetric for both host and plugin?

I think that can also work without lifetime management, yes. The main constraint here is not needing to keep track of lifetimes/specific open context menus. If the plugin can provide its own context menu items in the same was as the host could, and calling them follows the same 'invoke-menu-id-for-param-id' API on both sides then that could also work.

I'd advice against doing this with JSON, because that would mean that in order for plugins and hosts to use this they would need to add a JSON parsing library, and also recreate the data structures correctly from the parsed JSON.

**Comment by Bremmers** [comment]
> ```c
> ```c
> struct clap_menu_builder {
>    void add_entry(icon, label, entry_id);
>    void add_separator();
>    void begin_submenu(icon, label);
>    void end_submenu();
> };
> 
> struct clap_host_context_menu {
>    bool create(context, builder);
>    void clicked(entry_id);
> };
> ```

 Here it might be hard for the host to connect the entry_id back to an action, maybe entry_id could be replaced by an arbitrary blob of bytes that the plugin would copy?
> ```
I don't think that would be a problem in real-world applications. The host could use the lowest byte of entry_id for the index in the top level menu, and the next byte for the submenu index, for example.

**Comment by Bremmers** [comment]
If we'd go for a 'menu descriptor' instead of a 'menu builder' the plugin could pass a descriptor to the popup() function. And it could simply pass nil if it doesn't have anything.

**Comment by robbert-vdh** [comment]
Have a menu builder sounds like something we may want to avoid I'm very much in favor of just using a static data structure per parameter. It may also be possible to have this data structure be a simple flat list instead of needing to juggle 'multiple' menus. The reasoning behind this is that the plugin/host showing the menu will need to build its own context menu tree structure anyways, so presenting the input as a flat list shouldn't have many drawbacks.

The side of the interface that allows a plugin to integrate the host's context menu items would then look something like this:

```cpp
enum {
   CLAP_MENU_FLAG_DEACTIVATED = 1 << 0,
   CLAP_MENU_FLAG_IS_SEPARATOR = 1 << 1,
   CLAP_MENU_FLAG_CHECKED = 1 << 2,
};

/**
 * Describes a context menu item. Depending on the value of `parent_id`, this
 * context menu item may be a top level item or it may be part of a submenu.
 */
typedef struct clap_menu_item {
   char     name[CLAP_NAME_SIZE];
   uint32_t flags;
   /**
    * If this item is part of a submenu, then this is the linear index of the
    * parent menu item. Otherwise `CLAP_INVALID_ID`/`UINT32_MAX` indicates that
    * this menu item belongs to the top level.
    */
   uint32_t parent_id;
} clap_menu_item_t;

typedef struct clap_host_plugin_context_menu {
   /**
    * Get the number of host context menu items for a parameter.
    */
   uint32_t (*count)(const clap_host_t *host, clap_id param_id);

   /**
    * Get the descriptor for a host context menu item. These menu items remain
    * the same for the life of the parameter.
    */
   bool (*get_item)(const clap_host_t *host,
                    clap_id            param_id,
                    uint32_t           index,
                    clap_menu_item    *menuitem);

   /**
    * Trigger one of the context menu items for a parameter by its index. This
    * should not be called for menu items that have associated child menu items,
    * or for menu items with the `CLAP_MENU_FLAG_DEACTIVATED` or
    * `CLAP_MENU_FLAG_IS_SEPARATOR` flags.
    */
   void (*trigger)(const clap_host_t *host, clap_id param_id, uint32_t index);
} clap_host_plugin_context_menu_t;
```

**Comment by baconpaul** [comment]
Submenus are super important

**Comment by baconpaul** [comment]
And reconstructing from the index tree is backwards from now both vstgui and juce menu makers work (sorry pressed send too soon).

**Comment by Bremmers** [comment]
@robbert-vdh Not sure if your clap_menu_item works with nested submenus (a submenu in a submenu)?

**Comment by Bremmers** [comment]
Ah, nice.

This is my code with your flat list.
There's a problem with it though: the first parameter if clap_menu_descriptor should be 'plugin' if it's passed to clap_host_context_menu. But it's 'host' for clap_host_context_menu.getdescriptor.

```c
static CLAP_CONSTEXPR const char CLAP_EXT_CONTEXT_MENU[] = "clap.context-menu.draft/0";

enum {
  CLAP_MENU_FLAG_HASSUBMENU = 1 << 0,
  CLAP_MENU_FLAG_ISSEPARATOR = 1 << 1,
  CLAP_MENU_FLAG_DEACTIVATED = 1 << 2,
  CLAP_MENU_FLAG_CHECKED = 1 << 3,
  CLAP_MENU_FLAG_SELECTED = 1 << 4,   //'radio'
}

typedef struct clap_menu_item {
  char name[CLAP_NAME_SIZE];
  uint32_t flags;
  /**
   * If this item is part of a submenu, then this is the linear index of the
   * parent menu item. Otherwise `CLAP_INVALID_ID`/`UINT32_MAX` indicates that
   * this menu item belongs to the top level.
   */
  uint32_t parent_id;
}

typedef struct clap_menu_descriptor {
   /**
    * Get the number of context menu items for a parameter.
    */
   uint32_t (*count)(const clap_host_t *host, clap_id param_id);

   /**
    * Get the descriptor for a context menu item. These menu items remain
    * the same for the life of the parameter.
    */
   bool (*get_item)(const clap_host_t *host,
                    clap_id            param_id,
                    uint32_t           index,
                    clap_menu_item    *menuitem);
}

typedef struct clap_plugin_context_menu {
  //called on clicking an item from the clap_host_context_menu.popup() desciptor
  void (*clicked)(const clap_host_t *host, clap_id param_id, uint32_t index);
}

typedef struct clap_host_context_menu {
  bool (*getdescriptor)(const clap_host_t *host, clap_id param_id, clap_menu_descriptor *desciptor);

  // called on clicking a menu item
  // clicked() isn't called for menu items that have CLAP_MENU_FLAG_HASSUBMENU, CLAP_MENU_FLAG_ISSEPARATOR or CLAP_MENU_FLAG_DEACTIVATED in flags
  void (*clicked)(const clap_host_t *host, clap_id param_id, uint32_t index);

  //plugins that don't have menus can use popup() to let the host show a menu.
  //the plugin can pass a optional clap_menu_desciptor (or nil)
  void (*popup)(const clap_host_t *host, clap_id param_id, clap_menu_desciptor *descriptor);
}
```

**Comment by Bremmers** [comment]
Yeah, that would even look nice. Will do that later. I should take off my object-oriented hat I guess ;-)

**Comment by baconpaul** [comment]
I think using linear index as as the menu id is a mistake and we should not do it. Three reasons for this

1. Everywhere *else* in the CLAP API we have worked to separate linear index from id. params, ports, etc... and we should continue that
2. a linear view of a tree like menu is now how most folks write their UI so every host would have to flatten, and every plugin would have to unflatten, this data structure
3. A linear list of the span of the entire menu and submenu forbids the possibility of a very large and dynamic menu. Imagine a menu over the file system for instance. You would implicitly do a recursive find to answer the `count` API point

So I think a far better approach is to have a `menu_id` and require callbacks for submenus relative to their parent, rather than the opposite. So

```cpp
typedef struct clap_menu_item {
  char name[CLAP_NAME_SIZE];
  uint32_t flags;
 
  /*
   * Each menu item has a unique ID. The root menu item is named CLAP_CONTEXT_MENU_ITEM_ROOT
   */
  uint32_t menu_id;
}

typedef struct clap_host_context_menu {
  /**
   * Get the number of plugin context menu items for a parameter.
   */
  uint32_t (*count)(const clap_plugin_t *plugin, clap_id param_id, clap_id menu_id);

  /**
   * Get the description for a plugin context menu item. These menu items remain
   * the same for the life of the parameter.
   */
  bool (*get_item)(const clap_plugin_t *plugin,
                   clap_id            param_id,
                   clap_id            menu_id,
                   uint32_t          index,
                   clap_menu_item    *menuitem);

  void (*clicked)(const clap_plugin_t *plugin, clap_id param_id, clap_id menu_id);
}

```

then the act of constructing a submenu looks like a subsequent call to `count` for that items `menu_id`

**Comment by Bremmers** [comment]
Haha, we went full-circle today.
That would be something like this then:

```c
static CLAP_CONSTEXPR const char CLAP_EXT_CONTEXT_MENU[] = "clap.context-menu.draft/0";

enum {
  CLAP_MENU_FLAG_HASSUBMENU = 1 << 0,
  CLAP_MENU_FLAG_ISSEPARATOR = 1 << 1,
  CLAP_MENU_FLAG_DEACTIVATED = 1 << 2,
  CLAP_MENU_FLAG_CHECKED = 1 << 3,
  CLAP_MENU_FLAG_SELECTED = 1 << 4,   //'radio'
}

static const CLAP_CONSTEXPR clap_id CLAP_EXT_CONTEXT_MENU_ROOT = CLAP_INVALID_ID - 1;

typedef struct clap_menu_item {
  char name[CLAP_NAME_SIZE];
  uint32_t flags;
  /*
   * Each menu item has a unique ID. The root menu item is named CLAP_CONTEXT_MENU_ITEM_ROOT
   * CLAP_MENU_FLAG_HASSUBMENU in flags means there's a submenu. count/get_item are called with menu_item_id as menu_id parameter
   */
  clap_id menu_item_id;
}

typedef struct clap_plugin_context_menu {
  /**
   * Get the number of plugin context menu items for a parameter.
   */
  uint32_t (*count)(const clap_plugin_t *plugin, clap_id param_id, clap_id menu_id);

  /**
   * Get the description for a plugin context menu item.
   */
  bool (*get_item)(const clap_plugin_t *plugin,
                   clap_id            param_id,
                   clap_id            menu_id,
                   uint32_t          index,
                   clap_menu_item    *menuitem);

  // called on clicking a menu item
  // clicked() isn't called for menu items that have CLAP_MENU_FLAG_HASSUBMENU, CLAP_MENU_FLAG_ISSEPARATOR or CLAP_MENU_FLAG_DEACTIVATED in flags
  void (*clicked)(const clap_plugin_t *plugin, clap_id param_id, clap_id menu_id);
}

typedef struct clap_host_context_menu {
  /**
   * Get the number of host context menu items for a parameter.
   */
  uint32_t (*count)(const clap_host_t *host, clap_id param_id, clap_id menu_id);

  /**
   * Get the description for a host context menu item.
   */
  bool (*get_item)(const clap_host_t *host,
                   clap_id            param_id,
                   clap_id            menu_id,
                   uint32_t          index,
                   clap_menu_item    *menuitem);

  // called on clicking a menu item
  // clicked() isn't called for menu items that have CLAP_MENU_FLAG_HASSUBMENU, CLAP_MENU_FLAG_ISSEPARATOR or CLAP_MENU_FLAG_DEACTIVATED in flags
  void (*clicked)(const clap_host_t *host, clap_id param_id, clap_id menu_id);

  //plugins that don't have menus can use popup() to let the host show a menu.
  //the host will check clap_plugin_context_menu to see if the plugin has any additional menu items for the menu
  void (*popup)(const clap_host_t *host, clap_id param_id);
}
```

**Comment by baconpaul** [comment]
yeah that looks good to me.

**Comment by abique** [comment]
`void (*popup)(const clap_host_t *host, clap_id param_id);`
I'm not sure about that one, it asks the host to become a toolkit.

If you remove it, then the interface would be the same for both plugin and host, so you could just have one struct then: `clap_context_menu_provider`.

I also have the impression that for anything a bit more dynamic and complex the array access is less convenient than the menu builder approach. Because:

```C
uint32_t menu_count(const clap_host_t *host, clap_id param_id, clap_id menu_id)
{
   auto menu = compute_menu(param_id, menu_id);
   return menu.size();
}

bool menu_get_item(const clap_host_t *host,
                   clap_id            param_id,
                   clap_id            menu_id,
                   uint32_t          index,
                   clap_menu_item    *menuitem)

{
   auto menu = compute_menu(param_id, menu_id);
   *menuitem = menu[index];
   return true;
}
```
Of course in this example menu could be cached, but the issue is that this whole menu construction will be done twice, and cached once with this approach, and you don't know when to release that cached menu. While having a menu builder would allow to build the data structure only once.

`CLAP_MENU_FLAG_DEACTIVATED`: I think in the GUI toolkit it is more common to see `enabled/disabled`.

Can a menu entry have a tooltip?

Can a menu entry have some formatting? Like bold or italic?

I also have the feeling that this king of gui topic tend to require new features over time and I'm worried that this extension won't live long until people complain because it lacks this or that.

We could also consider exchanging a json string for describing the menu entries. This will be maybe a bit boring but in the long run, it will be possible to improve the menu overtime without breaking the extension.

**Comment by abique** [comment]
Also I don't understand well what the menu id is about?

**Comment by robbert-vdh** [comment]
How would the plugin trigger the host's context menu for a parameter without the popup function?.

**Comment by robbert-vdh** [comment]
> Also I don't understand well what the menu id is about?

Menus can have submenus. So with this design if a menu item has the `CLAP_MENU_FLAG_HAS_SUBMENU` flag, then the host/plugin can use the `count()` and `get_item()` functions on that menu item's `menu_id` to get its children.

And I still don't think using JSON here would solve anything. If you change the schema, then that is still a backwards incompatible change. And requiring hosts and plugins that want to implement this feature to a) include a JSON parsing library, and b) parse the schema **correctly** (in practice that doesn't always happen). That would both increase the bar for adopting this feature as well as make it easier to misuse the API (through slightly different implementations of the schema).

**Comment by baconpaul** [comment]
My one comment on this is: the parameter menus in bitwig and reaper are the one place where our users notice a real positive from VST3 which they don't see in other formats. They are incredibly useful when merged with our menus. So I think this is important!

And submenus matter for sure. Reaper uses them

But both platforms support this today, only using the VST3 api of "give me the menu gui object" which we then have to walk to get the data to put it in our menu.

**Comment by abique** [comment]
> How would the plugin trigger the host's context menu for a parameter without the popup function?.

The plugin would not, it is the right click that triggers it.

> > Also I don't understand well what the menu id is about?
> 
> Menus can have submenus. So with this design if a menu item has the `CLAP_MENU_FLAG_HAS_SUBMENU` flag, then the host/plugin can use the `count()` and `get_item()` functions on that menu item's `menu_id` to get its children.

Then I think it is even more confusing, if you do it like that then you should flatten the whole tree in a list and add `parent_item_id` in `clap_menu_item`.

> And I still don't think using JSON here would solve anything. If you change the schema, then that is still a backwards incompatible change. And requiring hosts and plugins that want to implement this feature to a) include a JSON parsing library, and b) parse the schema **correctly** (in practice that doesn't always happen). That would both increase the bar for adopting this feature as well as make it easier to misuse the API (through slightly different implementations of the schema).

You can extend JSON by adding new keys to dictionaries without breaking backward compatibility.

```C
  // called on clicking a menu item
  // clicked() isn't called for menu items that have CLAP_MENU_FLAG_HASSUBMENU, CLAP_MENU_FLAG_ISSEPARATOR or CLAP_MENU_FLAG_DEACTIVATED in flags
  void (*clicked)(const clap_host_t *host, clap_id param_id, clap_id menu_id);
```
Clicked should take a `menu_item_id` instead?

I'm still convinced that the menu builder is more convenient than an a possibly computed array iteration.

**Comment by robbert-vdh** [comment]
> The plugin would not, it is the right click that triggers it.

How? Does the host always listen for right clicks? How does the host know when a right click should pop up the host's own context menu? And how does the host know if the cursor is hovering over a parameter that should open a context menu when you right click on it?

> Then I think it is even more confusing, if you do it like that then you should flatten the whole tree in a list and add parent_item_id in clap_menu_item.

The reasoning for having these opaque menu IDs rather than a flat list is that you end up with a similar structure you'd need to build the context menu in most APIs. After calling `count()` and `get_item()` a couple of times, you essentially end up with a hash table mapping menu IDs (`clap_id`) to menus (lists of `clap_menu_item`). You can then traverse that menu starting at `CLAP_EXT_CONTEXT_MENU_ROOT` and then retrieving each child menu from the hash table as you encounter them. That's very simple to work with. I wouldn't be opposed to a flat list where the other side needs to reconstruct it into a tree, but this approach does skip a step which is nice.

> You can extend JSON by adding new keys to dictionaries without breaking backward compatibility.

I still think using JSON (or XML, or any other text based data format that needs to be manually parsed first) is a bad idea for all of the reasons I already named, especially when it's not necessary.

> Clicked should take a `menu_item_id` instead?

Each menu item is a menu. That's why all menu items have a `menu_item_id`. The `count()` function gets you the _children_ of a menu. So if a menu has a submenu, it will have >1 children. If it doesn't, then `count()` will return 0.

> I'm still convinced that the menu builder is more convenient than an a possibly computed array iteration.

The proposed data structure is a simple static directed acyclic graph. The benefit here is that it is trivial to work with. Hiding the data structure behind a data structure means that to be able to use this interface as a host or a plugin, you also need to implement the builder yourself. And unlike the current data structure, the builder could have dynamic behavior. This reminds me a lot of the LV2 atom forge, and I'd prefer to stick with static data structures if possible. Doing so would also makes bridging simpler, as otherwise you'd need to bridge each and every function call for this builder interface.

**Comment by abique** [comment]
> > The plugin would not, it is the right click that triggers it.
> 
> How? Does the host always listen for right clicks? How does the host know when a right click should pop up the host's own context menu? And how does the host know if the cursor is hovering over a parameter that should open a context menu when you right click on it?

Well when the host receives a right click, it shows the popup menu, and when the plugin receives a right then it is the plugin which shows the plugin menu.

**Comment by robbert-vdh** [comment]
How does the host know when it should handle that right click and when it should do nothing? And how does it know for what parameter it should show a context menu?

**Comment by Bremmers** [comment]
Robbert beat me to it, I'll post my replies anyway:

> The plugin would not, it is the right click that triggers it.
The plugin calls popup() when the user rightclicks a knob in the plugin UI. This is for plugins that don't have their own menus, so the host can provide a menu for the parameter associated with the knob.

> Clicked should take a menu_item_id instead?
It does! :-)

> I'm still convinced that the menu builder is more convenient than an a possibly computed array iteration.
This extension should be super-simple to implement for plugins, so they'll actually do so. Hosts will implement it anyway because they want to add their menu items to the plugin menus.

A menu builder would have to be provided by the plugin. This is quite a bit of code, and it may not fit in well with some plugin menu code so it's would be even more code.
Some kind of interator is easier for the plugin side I think. I'm totally open to any awesome iterator type ;-)

For the same reason JSON isn't a good idea I think. Virtually no plugin - or host, for that matter ;-) - has JSON in its code. It would have to add a parser just to add this extension. That doesn't make it easy to implement for plugins.


Perhaps we could add an additional 'more' field to clap_menu_item that allows for adding things like icons / formatting in the future. Or it could be a function like clap_host_context_menu.get_additional_info(host, param_id, menu_item_id, ...)


I've removed clap_plugin_context_menu which probably was a silly idea:
```c
//using this extension the host can add menu items to a plugin's parameter context menu
//a plugin that doesn't have context menus can call clap_host_context_menu.popup() so the host can show its own menu


static CLAP_CONSTEXPR const char CLAP_EXT_CONTEXT_MENU[] = "clap.context-menu.draft/0";

enum {
  CLAP_MENU_FLAG_HASSUBMENU = 1 << 0,
  CLAP_MENU_FLAG_ISSEPARATOR = 1 << 1,
  CLAP_MENU_FLAG_DEACTIVATED = 1 << 2,
  CLAP_MENU_FLAG_CHECKED = 1 << 3,
  CLAP_MENU_FLAG_SELECTED = 1 << 4,   //'radio'
}

static const CLAP_CONSTEXPR clap_id CLAP_EXT_CONTEXT_MENU_ROOT = CLAP_INVALID_ID - 1;

typedef struct clap_menu_item {
  char name[CLAP_NAME_SIZE];
  uint32_t flags;
  /*
   * Each menu item has a unique ID. The root menu item is named CLAP_CONTEXT_MENU_ITEM_ROOT
   * CLAP_MENU_FLAG_HASSUBMENU in flags means there's a submenu. count/get_item are called with menu_item_id as menu_id parameter
   */
  clap_id menu_item_id;
}

typedef struct clap_host_context_menu {
  // Get the number of host context menu items for a parameter.
  uint32_t (*count)(const clap_host_t *host, clap_id param_id, clap_id menu_id);

  // Get the description for a host context menu item.
  bool (*get_item)(const clap_host_t *host,
                   clap_id            param_id,
                   clap_id            menu_id,
                   uint32_t          index,
                   clap_menu_item    *menuitem);

  // called on clicking a host context menu item
  void (*clicked)(const clap_host_t *host, clap_id param_id, clap_id menu_item_id);

  //plugins that don't have menus can use popup() to let the host show a menu.
  void (*popup)(const clap_host_t *host, clap_id param_id);
}
```

**Comment by abique** [comment]
> How does the host know when it should handle that right click and when it should do nothing? And how does it know for what parameter it should show a context menu?

If the right click happens on the host gui (@baconpaul said it was the most useful scenario), then the host will show a popup menu and request additional items from the plugins and inserts them into the popup menu.

If the right click happens on the plugin gui, then the plugin will show a popup menu and request additional items from the host.

Anyway, I step out of the discussion.
I let you guys figure it out and make a proposal in the form of a pull request towards the next branch, with clear explanation for:
- what problem this extension solves and why it is useful
- concrete usage examples
- example code for the implementation

Does it sound good?

**Comment by abique** [comment]
> > The plugin would not, it is the right click that triggers it.
> 
> The plugin calls popup() when the user rightclicks a knob in the plugin UI. This is for plugins that don't have their own menus, so the host can provide a menu for the parameter associated with the knob.

And the host isn't a GUI toolkit, if the plugins wants to show a popup menu it should solves this problem itself.
